### PR TITLE
rgw:[bug] fix data consistency error casued by rgw sent timeout

### PR DESCRIPTION
**Pull request by ligangbin117**
Incomplete data transmission and incorrect data re-transmission

Signed-off-by: 李纲彬82225 <82225@sangfor.com>

### CODE REVIEW

File: src/rgw/rgw_civetweb.cc [hunk 1]
```diff
@@ -9,22 +9,31 @@
...
+#define dout_context g_ceph_context
 #define dout_subsys ceph_subsys_rgw
 
 size_t RGWCivetWeb::write_data(const char *buf, const size_t len)
 {
+  size_t off = 0;
   auto to_sent = len;
   while (to_sent) {
-    const int ret = mg_write(conn, buf, len);
+    const int ret = mg_write(conn, buf + off, to_sent);
```
Comment (cbodley) [inline src/rgw/rgw_civetweb.cc]: looks right to me, but can we drop the dout messages?

**Comment by cbodley** [comment]
@ligangbin117 do you have a way to reproduce this issue against master? looking at `mg_write()` in civetweb, it calls a `push_all()` function that loops until it's sent all of the data. so i don't this change is necessary on master, but it may apply to earlier releases

File: src/rgw/rgw_civetweb.cc (hunk 1 above)
Comment (ligangbin117) [reply src/rgw/rgw_civetweb.cc]: ok

**Comment by zhangsw** [comment]
@ligangbin117 Hi, could you reproduce this issue?

**Comment by ligangbin117** [comment]
This is the latest code for the civetweb project, there has a timeout and break in push_all().
If the network environment is not good, there may be timeout interruption and incomplete transmission

static int
push_all(struct mg_context *ctx,
         FILE *fp,
         SOCKET sock,
         SSL *ssl,
         const char *buf,
         int len)
{
	double timeout = -1.0;
	int n, nwritten = 0;

	if (ctx == NULL) {
		return -1;
	}

	if (ctx->dd.config[REQUEST_TIMEOUT]) {
		timeout = atoi(ctx->dd.config[REQUEST_TIMEOUT]) / 1000.0;
	}

	while ((len > 0) && (ctx->stop_flag == 0)) {
		**n = push_inner(ctx, fp, sock, ssl, buf + nwritten, len, timeout);**
		if (n < 0) {
			if (nwritten == 0) {
				nwritten = -1; /* Propagate the error */
			}
			**break;**
		} else if (n == 0) {
			break; /* No more data to write */
		} else {
			nwritten += n;
			len -= n;
		}
	}

	return nwritten;
}

**Comment by joke-lee** [comment]
@cbodley here is my  way to reproduce this issue 

this is a file 139MiB size named 201910101336581608-201909121510598784-CMCC_CN_P855A23V1.0.0B18_TO_CMCC_CN_P855A23V1.0.0B19.zip in bucket test 

set the timeout to 1s 
```
rgw frontends = "civetweb port=7480 request_timeout_ms=1000"
```

restart rgw

and download file with follow script

```
import socket,hashlib
host = '192.168.198.137'
uri = '/test/201910101336581608-201909121510598784-CMCC_CN_P855A23V1.0.0B18_TO_CMCC_CN_P855A23V1.0.0B19.zip?AWSAccessKeyId=yly&Expires=2203897620&Signature=FUDeErIP2JHOEHNRE42eMflfIKQ%3D'
mysock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
mysock.connect((host, 7480))

domain = host +":7480"
mysock.send('GET %s HTTP/1.0\r\nHost:%s\r\nAccept-Encoding:gzip,deflate\r\nUser-Agent:yuliyangtest13\r\nRange:bytes=0-\r\n\r\n' % (uri, domain))
count = 0
all_data = ""
index = 0
while True:
    data = mysock.recv(5120)
    if ( len(data) < 1 ) : break
    # time.sleep(0.25)
    count = count + len(data)
    print len(data), "%sMiB" % (count/1024/1024 , )
    all_data = all_data + data

mysock.close()
# Look for the end of the header (2 CRLF)
pos = all_data.find("\r\n\r\n");
print 'Header length',pos
print all_data[:pos]
# Skip past the header and save the picture data
body = all_data[pos+4:]

with open("download3.zip", "wb") as f:
    f.write(body)
```

you will see the all_data is growing large than 139M

image

**Comment by cbodley** [comment]
i opened https://tracker.ceph.com/issues/42816 to track backports

**Comment by yuriw** [comment]
[wip-yuri8-testing-2019-12-11-2105](https://trello.com/c/QPjkDkCb/778-wip-yuri8-testing-2019-12-11-2105)

**Pull request merged**